src/compiler/parser.rs

```diff
@@ -246,17 +335,116 @@ impl ParserType {
         }
 
         match self.token() {
-            _ => self.is_identifier(),
+            _ => {
+                if self.is_binary_operator() {
+                    return true;
+                }
+
+                self.is_identifier()
+            }
+        }
+    }
+
+    fn parse_expression(&mut self) -> Expression {
+        let expr = self.parse_assignment_expression_or_higher();
+
+        expr
+    }
+
+    fn parse_assignment_expression_or_higher(&mut self) -> Expression {
+        let expr = self.parse_binary_expression_or_higher(OperatorPrecedence::Lowest);
+
+        self.parse_conditional_expression_rest(expr)
+    }
+
+    fn parse_conditional_expression_rest(&self, left_operand: Expression) -> Expression {
+        left_operand
+    }
+
+    fn parse_binary_expression_or_higher(&mut self, precedence: OperatorPrecedence) -> Expression {
+        let left_operand = self.parse_unary_expression_or_higher();
+        self.parse_binary_expression_rest(precedence, left_operand)
+    }
+
+    fn parse_binary_expression_rest(
+        &self,
+        precedence: OperatorPrecedence,
+        left_operand: Expression,
+    ) -> Expression {
+        loop {
+            let new_precedence = get_binary_operator_precedence(self.token());
+
+            let consume_current_operator = new_precedence > precedence;
+
+            if !consume_current_operator {
+                break;
+            }
         }
+
+        left_operand
     }
 
-    // fn parse_expression(&self) -> Box<dyn Expression> {}
+    fn is_binary_operator(&self) -> bool {
+        get_binary_operator_precedence(self.token()) > OperatorPrecedence::Comma
```

Comment:
It'd be much more clear to implement this on `OperatorPrecendence` itself, I think.

Reply:
Ya in the Typescript code it's actually written as `> 0` since I guess you can mix/match number-valued enums + actual numbers, but so here I'm explicitly referencing the equivalent enum variant (which I think is actually clearer anyway). Ya basically `getBinaryOperatorPrecedence()` either returns a precedence that's higher than `OperatorPrecedence.Comma` or returns `-1` (ie `OperatorPrecedence.Invalid`) so does seem equivalent to just saying "is the current token in the set of binary operators?"